### PR TITLE
Aggregates snapshot

### PR DESCRIPTION
**Pull request by fabriziosestito**
Enables aggregate snapshots, see: https://hexdocs.pm/commanded/aggregates.html#aggregate-state-snapshots

To do this, I needed to refactor the aggregate state to use the `deftype` macro.

### CODE REVIEW

**Review by arbulu89 — COMMENTED** [review]
Why did you need to change to use `deftype`? In the example of the refence as see `defstruct` (which was our old implementation).
Is it for the polymorphich fields?

**Comment by fabriziosestito** [comment]
> Why did you need to change to use `deftype`? In the example of the refence as see `defstruct` (which was our old implementation). Is it for the polymorphich fields?

@arbulu89 it is used to cast the snapshot saved in the db to the aggregate state as we do with events. the example assumes you have already some sort of casting in place.

**Review by CDimonaco — APPROVED** [review]

File: lib/trento/domain/sap_system/instance.ex [hunk 1]
```diff
@@ -3,25 +3,19 @@ defmodule Trento.Domain.SapSystem.Instance do
...
-          host_id: String.t(),
-          health: Health.t(),
-          system_replication: String.t(),
-          system_replication_status: String.t()
-        }
+  use Trento.Type
+
+  deftype do
+    field :sid, :string
+    field :instance_number, :integer
```
Comment (arbulu89) [inline lib/trento/domain/sap_system/instance.ex:14]: I'm not sure if we should change the instance number to integer. Yes it is an integer, but we always use it as a 2 digit string. I would be conservative with this an keep string

File: lib/trento/domain/sap_system/instance.ex [hunk 2]
```diff
@@ -3,25 +3,19 @@ defmodule Trento.Domain.SapSystem.Instance do
...
+    field :features, {:array, :string}
+    field :host_id, :string
```
Comment (arbulu89) [inline lib/trento/domain/sap_system/instance.ex:16]: `host_id` maybe as UUID?

**Review by arbulu89 — APPROVED** [review]
Good!
Commented some minor things that you could change without other review

File: lib/trento/domain/sap_system/instance.ex (hunk 1 above)
Comment (fabriziosestito) [reply lib/trento/domain/sap_system/instance.ex:14]: mmh i think i mistyped it thank you

**Review by dottorblaster — APPROVED** [review]
LGTM

**Review by nelsonkopliku — APPROVED** [review]
Nice! Thanks!

**Pull request merged**